src/algebra/continued_fractions/computation/correctness_terminating.lean

```diff
@@ -0,0 +1,217 @@
+
+## Main Theorems
+
+- `gcf.comp_exact_value_correctness_of_stream_eq_some` shows that `gcf.comp_exact_value` indeed
+  returns the value `v` when given the convergent and fractional part as described in the summary.
+- `gcf.of_correctness_of_terminated_at` shows the equality `v = (gcf.of v).convergents n`
+  if `gcf.of v` terminated at position `n`.
+-/
+
+namespace generalized_continued_fraction
+open generalized_continued_fraction as gcf
+
+variables {K : Type*} [discrete_linear_ordered_field K] {v : K} {n : ℕ}
+
+/--
+Given two successive continuants (at positions `n + 1` and `n`) of a continued fraction
+`generalized_continued_fraction.of v` and the fractional part at `int_fract_pair.stream n`,
+this function returns the exact value approximated by the continued fraction, that is `v`.
+
+For example, let `(v : ℚ) := 3.4`. We have:
+  `gcf.int_fract_pair.stream v 0 = some ⟨3, 0.4⟩`, and
+  `gcf.int_fract_pair.stream v 1 = some ⟨2, 0.5⟩`,
+
+Now `(gcf.of v).convergents 1 = 3 + 1/2`, and our fractional term (the residual) at position `2` is
+`0.5`. We hence have `v = 3 + 1/(2 + 0.5) = 3 + 1/2.5 = 3.4`.
+-/
+protected def comp_exact_value (pred_conts conts : gcf.pair K) (fr : K) : K :=
+-- if the fractional part is zero, we exactly approximated the value by the last continuants
+if fr = 0 then conts.a / conts.b
+-- otherwise, we have to include the fractional part in a final continuants step.
+else let exact_conts := next_continuants 1 fr⁻¹ pred_conts conts in
+  exact_conts.a / exact_conts.b
+
+variable [floor_ring K]
+
+/-- Just a computational lemma we need for the next main proof. -/
+protected lemma comp_exact_value_correctness_of_stream_eq_some_aux_comp {a : K} (b c : K)
+  (fract_a_ne_zero : fract a ≠ 0) :
+  ((⌊a⌋ : K) * b + c) / (fract a) + b = (b * a + c) / fract a :=
+begin
+  have : ((⌊a⌋ : K) * b + c) / (fract a) + b = ((b * (fract a) + b * ⌊a⌋ + c) / fract a), by
+  { field_simp [fract_a_ne_zero], ac_refl },
+  simpa [←left_distrib] using this
+end
+
+open generalized_continued_fraction as gcf
+
+/--
+Shows the correctness of `comp_exact_value` in case the continued fraction did not terminate.
+That is, at every step of the computation of a continued fraction `gcf.of v`, we obtain the value
+`v` if we compute the continuant at position `n` and add the residual term given by the fractional
+part stored in `int_fract_pair.stream v n`.
+-/
+lemma comp_exact_value_correctness_of_stream_eq_some :
+  ∀ {ifp_n : int_fract_pair K}, int_fract_pair.stream v n = some ifp_n →
+    v = gcf.comp_exact_value
+          ((gcf.of v).continuants_aux  n)
+          ((gcf.of v).continuants_aux $ n + 1)
+          ifp_n.fr :=
+begin
+  let g := gcf.of v,
+  induction n with n IH,
+  { assume ifp_zero stream_zero_eq, -- nat.zero
+    have : int_fract_pair.of v = ifp_zero, by
+    { have : int_fract_pair.stream v 0 = some (int_fract_pair.of v), from rfl,
+      simpa only [this] using stream_zero_eq },
+    cases this with refl,
```

Comment:
```suggestion
    cases this,
```
Casing on an equality doesn't produce any hypotheses. There's nothing to name `refl`. Note that this is not the same as `rcases ... with ...rfl...`.

Also in various places below.

Reply:
Thanks - I do not know why I had this here; must be some noise from a previous proof attempt. I fixed it.